pkgs/README.md

```diff
@@ -384,7 +384,7 @@ All versions of a package _must_ be included in `all-packages.nix` to make sure
 * `meta.license` must be set and match the upstream license.
   * If there is no upstream license, `meta.license` should default to `lib.licenses.unfree`.
   * If in doubt, try to contact the upstream developers for clarification.
-* `meta.mainProgram` must be set when appropriate.
+* `meta.mainProgram` must be set when there is any main executable for the package in `$bin/bin` or `$out/bin`. Always use a hardcoded string (don't use `pname`, for example).
```

Comment:
"there is any main executable"

@yu-re-ka @infinisil @Atemu  would we agree that main executable for coreutils is `"coreutils"`, imagemagic is `"magick"` busybox is `"busybox"` ?

anything that doesn't fit this ?

Reply:
What about `xfsprogs` or `dosfstools`?

Reply:
@yu-re-ka those are perfect examples

I think for me at least they obviously don't pass "there is any main executable"

what do we do about coreutils, imagemagick, busybox?

Reply:
I wouldn't set a mainProgram for those either.

Reply:
I would not agree with that and I'd prefer a more elaborate wording here aswell as a link to the discussion in https://github.com/NixOS/nixpkgs/issues/281276 for additional context.

IMO, only packages which contain an executable which could be considered the *primary way of interacting with the package* should receive a `meta.mainProgram`.

***

I've literally never ran the `coreutils` command. It is not how I interact with the `pkgs.coreutils` package. I didn't even know it existed.  
I think `pkgs.coreutils` is pretty much the antithesis for `meta.mainProgram` as I'd consider almost all of its executables to be worthy of their own package.

I would not generally exclude packages with multiple executables from `meta.mainProgram` though as it's often the case that one of them is much more likely to be used than others.  
A good example of this would be java JDKs/JREs whose `meta.mainProgram` should obviously be `java` but contain a bunch of other java-related tools aswell.  
Another would be `pkgs.xz` which has both `xz` and `unxz` which could be considered equally useful but I'd prefer the former as it's a tool for compression.

The only way I've ever interacted with the fs progs are their `mkfs` commands, so I could see those being the `meta.mainProgram` but I could also see the case against it because they often also contain other similarly useful commands (though I rarely use them in comparison).

Magick is a lot more nuanced. I could see the case for making `convert` its `meta.mainProgram` as that's the only way I've ever used it but I imagine there's lots of other ways people would commonly want to use it.  
I wouldn't set a `meta.mainProgram` unless I had some (anec)data that most others are in the same boat as me.

I could however see the case for `pkgs.busybox` as its `busybox` command can be used to run all the other commands like this: `busybox ls`. Eventhough the `busybox` executable isn't the primary way I'd interact with it if I had it installed permanently, `nix run nixpkgs#busybox ls`/`"${lib.getExe pkgs.busybox} ls"` seem quite idiomatic and useful to me.  
It's a bit of a special case though.

***

I think a good question to ask yourself here is: `If I needed to install this package because I need to run one program, which program would that be?"  
If there is one program that is much more desirable or frequently used than all other programs the package provides, that should be the package's `meta.mainProgram`.  
If there are >1 answers with similar desirability or popularity, the package shouldn't have a `meta.mainProgram`. (Or perhaps we should add an explicit marker for this case? `meta.mainProgram = null`?

That should provide a rather clear answer for one person but the remaining issue is that the answer can be different for multiple people.  
This I would solve by trying to answer the question for the general public who will use the package rather than yourself. Here it could be helpful to ask yourself whether your use-case is common or niche. There may be someone whose main way of interacting with Java JDKs is to manually invoke `javac` for instance but most people would not expect that to be its `meta.mainProgram`.

An easy way to do this would be to simply ask in the PR adding the package and/or `meta.mainProgram`. If the opinion is split and it's not just someone having a super niche use-case, there should probably be no `meta.mainProgram`.

Reply:
(I do not consider this resolved.)

Reply:
@Atemu I really like "considered the primary way of interacting with the package" maybe we can use this wording in documentation 

I also like your last paragraph . Maybe case by case treatment is easier 

> easy way to do this would be to simply ask in the PR adding the package and/or meta.mainProgram. If the opinion is split and it's not just someone having a super niche use-case, there should probably be no meta.mainProgram.

Reply:
My feedback on the PR is what I wrote in the first paragraph. It's still rather ambiguous.

The rest of the comment offers discussion on how we could achieve that; a better defined purpose of `meta.mainProgram`.

Reply:
@kirillrdy 

Perhaps it could be worded a little something like this:

> In case you are unsure whether there should be a `meta.mainProgram` or which program it should be, it can be helpful to reach out to reviewers in your PR. If there is a split opinion on the specific program with common use-cases on either side, there should probably be no `meta.mainProgram`.

Reply:
@Atemu thank you for your input.I am ok with this PR as it is.

Reply:
@teto whats wrong with nix run nixpkgs#coretuils running $out/bin/coretuills it was already designed as proxy ?

Reply:
Let's keep this on topic. @Atemu do you have any specific feedback remaining?

Reply:
coreutils is a wrong example then, I was thinking of packages that shouldn't have a mainProgram. Right now if you dont have a `mainProgram`, this feels like an error instead of being a deliberate choice.